### ISSUE TITLE
kubelet fails to heartbeat with API server with stuck TCP connections

### ISSUE THREAD

**Issue by derekwaynecarr** [kubernetes/kubernetes#48638]
**Is this a BUG REPORT or FEATURE REQUEST?**:
/kind bug

**What happened**:
operator is running an HA master setup with a LB in front.  kubelet attempts to update node status, but `tryUpdateNodeStatus` wedges.  based on the goroutine dump, the wedge happens when it attempts to GET the latest state of the node from the master.  operator observed 15 minute intervals between attempts to update node status when kubelet could not contact master.  assume this is when the LB ultimately closes the connection.  the impact is that node controller then marked node as lost, and workload was evicted.

**What you expected to happen**:
expected the kubelet to timeout client-side.
right now, no kubelet->master communication has a timeout.
ideally, the kubelet -> master communication would have a timeout based on the configuration of the node-status-update-frequency so that no single attempt to update status wedges future attempts.

**How to reproduce it (as minimally and precisely as possible)**:
see above.

**Comment by vishh** [comment]
cc @wojtek-t @gmarek

**Comment by derekwaynecarr** [comment]
we have to setup some timeout, but i am worried about the potential impact on any watch code.

**Comment by derekwaynecarr** [comment]
note: this is different than the Dial timeout (which we do appear to have default).

**Comment by xiangpengzhao** [comment]
I remember @jayunit100 has a related PR but I don't know if it has been merged finally.

**Comment by derekwaynecarr** [comment]
need to take a pass at kube-proxy too.  i see no explicit timeout there as well at first glance.

**Comment by ncdc** [comment]
@derekwaynecarr fyi I reported this earlier as #44557

**Comment by gmarek** [comment]
#48926 is a blocker for this.

**Comment by derekwaynecarr** [comment]
@obeattie had a recommendation worth evaluating here:
https://github.com/kubernetes/kubernetes/issues/41916#issuecomment-315848607

it would require us to tune net.ipv4.tcp_retries2
see: https://www.kernel.org/doc/Documentation/networking/ip-sysctl.txt

in practice, it would make sense for this value to potentially be monitored by the kubelet to ensure node status update interval falls within desired range.

@vishh @gmarek - what are your thoughts?  longer term, we should still split client timeouts for long vs short operations, but this tweak would help address many situations where LB communication to a master introduces problems.

**Comment by obeattie** [comment]
@derekwaynecarr Have you seen #48670?

**Comment by derekwaynecarr** [comment]
@obeattie will take a look, looks promising.

**Comment by SleepyBrett** [comment]
We've had three major events in the last few weeks that comes down to this problem. Watches set up through an elb node that gets replaced or scaled down cause large numbers of nodes to go not ready for 15 minutes causing very scary cluster turbulence. ( We've generally seen between a third to half the nodes go not ready ). We're currently evaluating other ways to load balance the api servers for the components we currently send through the elb ( I haven't poured through everything but I think that boils down to the kubelet and the proxy (possibly flannel) ).

**Comment by gmarek** [comment]
I know it's not strictly related to this problem, but after a third of a 'zone' goes down NodeController will drastically reduce ratio in which it evicts Pods - exactly to protect you from screwing your cluster too much in situations like this.

Given that it is important problem you could try pushing @kubernetes/sig-api-machinery-feature-requests to make required changes to client-go

**Comment by jdumars** [comment]
/sig azure
Added for visibility

**Comment by vboginskey** [comment]
@derekwaynecarr @liggitt #52176 doesn't resolve this issue. See: https://github.com/kubernetes/kubernetes/pull/48670#issuecomment-352257836, https://github.com/kubernetes/kubernetes/pull/52176#issuecomment-349647651, and https://github.com/kubernetes/kubernetes/pull/52176#issuecomment-353319760.

**Comment by sstarcher** [comment]
@derekwaynecarr @liggitt Wanted to reiterate we are seeing this issue in k8s 1.9.3 
https://github.com/kubernetes/kops/issues/4946

**Comment by liggitt** [comment]
reopening, the hang is resolved but the underlying stuck TCP connection issue is not

**Comment by recollir** [comment]
I am wondering if this would also apply to the kube-proxy. As it also maintains a “connection” to the api server and would suffer from the same (???).

**Comment by redbaron** [comment]
Ideally fix should be in client-go

**Comment by 2rs2ts** [comment]
@recollir yes, I'm pretty sure you're right. This issue should be rescoped to include kube-proxy IMO

**Comment by liggitt** [comment]
one issue at a time :)

persistent kubelet heartbeat failure results in all workloads being evicted. kube-proxy network issues are disruptive for some workloads, but not necessarily all

kube-proxy (and general client-go support) would need a different mechanism, since those components do not heartbeat with the api like the kubelet does. I'd recommend spawning a separate issue for kube-proxy handling of this condition.

**Comment by obeattie** [comment]
Since this issue has been re-opened, would there be any value in me re-opening my PR for [this commit](https://github.com/monzo/kubernetes/commit/81f7a18a9adc09d2b9103148427f8feaa25ee984)? Monzo has been running this patch in production since last July and it has eliminated this problem entirely, for all uses of `client-go`.

**Comment by liggitt** [comment]
> Since this issue has been re-opened, would there be any value in me re-opening my PR for this commit? Monzo has been running this patch in production since last July and it has eliminated this problem entirely, for all uses of client-go.

I still have several reservations about that fix:
* it relies on behavior that appears to be undefined (it changes properties on the result of [`net.TCPConn#File()`](https://golang.org/pkg/net/#TCPConn.File), documented as "The returned os.File's file descriptor is different from the connection's. Attempting to change properties of the original using this duplicate may or may not have the desired effect.")
* calling `net.TCPConn#File()` has implications I'm unsure of: "On Unix systems this will cause the SetDeadline methods to stop working."
* it appears to only trigger closing in response to unacknowledged *outgoing* data... that doesn't seem like it would help clients (like kube-proxy) with long-running receive-only watch stream connections

that said, if @kubernetes/sig-api-machinery-pr-reviews and/or @kubernetes/sig-network-pr-reviews feel strongly that is the correct direction to pursue, that would be helpful to hear.

**Comment by redbaron** [comment]
Few notes on these very valid concerns:
 * https://golang.org/pkg/net/#TCPConn.File is returning dup'ed  filedescriptor, which AFAIK share all underneath structures in kernel, except for entry in file descriptor table, so either can be used with same results. Program should be aware not to try to use them simultaneously though, for exact same reasons.
 * today returned filedescriptor is set to blocking mode. Probably it can be mitigated by setting it back to nonblocking mode. In Go 1.11 returned fd is going to be in same blocking/unblocking mode as it was before .File() call: https://github.com/golang/go/issues/24942
 * Maybe it will not help simple watchers, I am not familiar with Informers internals, but I was under the impression that they are not only watching, but also periodically resyncing state, these resync would trigger outgoing data transfer which would then be detected .

**Comment by obeattie** [comment]
Indeed: as far as I understand, the behaviour is not _undefined_, it's just defined in Linux rather than in Go. I think the Go docs could be clearer on this. Here's the relevant section from `dup(2)`:

>After a successful return from one of these system calls, the old and new file descriptors may be used interchangeably. **They refer to the same open file description** (see `open(2)`) and thus share file offset and file status flags; for example, if the file offset is modified by using `lseek(2)` on one of the descriptors, the offset is also changed for the other.
>
>The two descriptors do not share file descriptor flags (the close-on-exec flag).

My code doesn't modify flags after obtaining the fd, instead its only use is in a call to `setsockopt(2)`. The docs for that call are fairly clear that it modifies properties of the socket _referred to_ by the descriptor, not the descriptor itself:

>`getsockopt()` and `setsockopt()` manipulate options for the socket **referred to by the file descriptor `sockfd`.** 

I agree that the original descriptor being set to blocking mode is annoying. Go's code is clear that this will not prevent anything from working, just that more OS threads may be required for I/O:

https://github.com/golang/go/blob/516f5ccf57560ed402cdae28a36e1dc9e81444c3/src/net/fd_unix.go#L313-L315

Given that a single Kubelet (or otherwise use of client-go) establishes a small number of long-lived connections to the apiservers, and that this will be fixed in Go 1.11, I don't think this is a significant issue.

I am happy for this to be fixed in another way, but given we know that this works and does not require invasive changes to the apiserver to achieve, I think it is a reasonable solution. I have heard from several production users of Kubernetes that this has bitten them in the same way it bit us.

**Comment by liggitt** [comment]
this issue is resolved for the kubelet in #63492

https://github.com/kubernetes/kubernetes/issues/65012 is open for the broader client-go issue

/close

**Comment by workhardcc** [comment]
Did scheduler & controller → api-server has the same issue?

**Comment by redbaron** [comment]
@workhardcc , yes, both use client-go , which AFAIK remains unfixed

**Comment by corest** [comment]
I'm experiencing this again with k8s 1.13 and azure only (filtered with `kubelet_node_status.go`):
```
I0319 17:14:25.579175   68746 kubelet_node_status.go:478] Setting node status at position 6
I0319 17:29:29.987934   68746 kubelet_node_status.go:478] Setting node status at position 7
```
For 15 minutes node doesn't update status. There are no issues on network level and node actually is fully operational. Adjusting `--node-monitor-grace-period=` on controller-manager helps, but that is not a solution

**Comment by svend** [comment]
@liggitt 

> this issue is resolved for the kubelet in #63492

I am running into this issue with Kubernetes 1.14.1. It looks like the fix in #63492 is in the kubelet client certificate code. Will this still work if kubelet is using token (not certificate) authentication?

**Comment by liggitt** [comment]
This regressed, and was refixed in 1.14.3

See https://github.com/kubernetes/kubernetes/pull/78016

**Comment by JohnRusk** [comment]
@derekwaynecarr , you wrote

>we have to setup some timeout, but i am worried about the potential impact on any watch code.

It turns out, there is an impact on watch code, but it doesn't happen with kubemark/hollow nodes, therefore it escaped detection for a long time.  

I've tagged you in on the PR I just made to fix the issue, and would appreciate your comments on whether that PR looks reasonable.
